### ISSUE TITLE
Log artifacts with step

### ISSUE THREAD

**Issue by fellnerse** [mlflow/mlflow#1554]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macos 10.14.5
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**:  1.0.0
- **Python version**:  3.7.3
- **Exact command to reproduce**: https://mlflow.org/docs/latest/python_api/mlflow.models.html#mlflow.models.Model.log

### Describe the problem
Logging a model with the mlflow Model.log function will use internally the log_artifact function which does not provide any possibility to provide a step like suggested here https://github.com/mlflow/mlflow/issues/32#issuecomment-461017625. Is there still current work going on in this direction? Logging a model without specifying the step/epoch/iteration it was generated doesn't make a lot of sense. For our project we would like to get the best model of a run/experiment where best means:
During the training some metrics and checkpoints/models are logged periodically (currently it all happens at the same time i.e. once per epoch). Now get the the run and timestamp when some logged metric is a global maximum/minimum (across all experiments, or all runs in one experiment). But at this point it gets hard because the models/artifacts are not mapped to the metrics (or the other way around). 

Is there something I'm missing out on? Is there some way how this is supposed to be done? I would say it's a crucial feature to get the best model for a run/experiment given some metric-key. 

Thank you in advance.

**Comment by timonbimon** [comment]
I am having the same problem! I want to find the best model over multiple runs, but since I can have multiple checkpoints over time in one run and the metrics (which are logged with the "step" in each run) seem not to be connected to the logged models (which are just connected to a run if I understand correctly?)

any idea on how to do this?

**Comment by dbczumar** [comment]
@fellnerse @timonbimon The `artifact_path` parameter of various `mlflow.*.log_model` commands is customizable. You can include step values in the `artifact_path`. Here's a simple example that logs a `pyfunc` model after each training iteration and embeds the iteration number ("step") in the artifact path:

```
import mlflow
import mlflow.pyfunc

import numpy as np


class TestModel(mlflow.pyfunc.PythonModel):

    def __init__(self, step):
        self.step = step

    def predict(self, context, model_input):
        return self.step


def train(step):
    return TestModel(step)


def evaluate_loss(model, data):
    return model.step * np.random.rand()


with mlflow.start_run():
    for step in range(100):
        model = train(step)
        loss = evaluate_loss(model=model, data=np.random.random(10))
        mlflow.log_metric("loss", loss, step=step)
        mlflow.pyfunc.log_model(
            python_model=model,
            artifact_path="python_model/{step}".format(step=step)
        )
```

We do not intend to add a `step` parameter to `log_artifact` at this time.

**Comment by timonbimon** [comment]
Thanks for your answer! :)

tbh I find it a little bit weird though, that to do one of the major things that MLflow Tracking (I assume) is designed for (namely tracking machine learning runs and then using those tracked runs to get the best trained model) has to be done with a hack?

Am I misunderstanding something in the design of MLflow or what it should be used for?
How are you supposed to use MLflow to find the best checkpoint for a deep learning model that is trained over many epochs?

**Comment by apurva-koti** [comment]
I understand your use case, but I'm not sure how logging artifacts with step solves your problem. For instance, logging metrics with step is obviously crucial; if I wanted to check the step with the lowest loss, I could look at the graph and find the global minimum. I can't do this with artifacts because they can't be numerically represented.
You could log the "best" model (as defined by some metric you want) by calling log_artifact accordingly in your training loop.

**Comment by timonbimon** [comment]
Yes, exactly! And after I looked at the loss vs step graph and found the step with the lowest loss, I will now want to of course get the checkpoint I saved for that step, since that corresponds to my best model I now might want to deploy.  

It's possible I have some huge misunderstanding here somewhere, so let me try to make my thinking even more explicit:

I am training a ResNet for an image classification task on a custom dataset. Just like Andrew Ng would want I have chose my single number evaluation metric. Now, after every epoch of training I will evaluate my current model on a validation set, log the metrics and if my the evaluation metric I care about is the best I have seen so far, I will want to save my model (which as I understand in MLflow would mean logging a checkpoint of the model as an artifact). Of course as I progress with training most likely there will be multiple epochs where the evaluation metric is the best I have seen so far (after the first epoch this will be my best model for sure, hopefully after the second epoch it will be even better), so I will end up with multiple checkpoints logged as I train for many epochs. 
Now at the end of the training I would like to find the checkpoint where my single evaluation metric did best and deploy that model to production. 

This seems like a pretty standard use-case to me (is it not? is logging all the best models a bad idea? how else would I do it?) and it is still not clear to me how MLflow supports this (since there seems no way to connect the checkpoints (artifacts) I logged to the value of the metric at a given step. Am I missing something? Should I not be saving my model checkpoints as artifacts? Is there another way to connect my model checkpoints with the metrics I am logging with MLflow).

Thanks already for your patience in going through this. :)

**Comment by apurva-koti** [comment]
I think the misunderstanding here is that a 'step' is not a snapshot of your training session at a given point of time, but rather the data used to relate metrics and parameters to time so they can be viewed graphically. Artifacts are semantically related with a _run_, not a step, so logging an artifact with step doesn't make a lot of sense to me. 
More simply, if your motivation is to log an artifact against the step that saw the lowest validation loss, you could achieve this with a combination of @dbczumar 's suggestion (which attaches the step to the artifact name) and keeping track of the lowest validation loss so far in your loop:
```python
if val_loss < lowest_val_loss:
    mlflow.log_artifact(...)
    val_loss = lowest_val_loss
```

**Comment by timonbimon** [comment]
I see, but then I think this is not really a misunderstanding then. I guess my point from the very beginning was that we seem to give up a lot of flexibility by only relating artifacts to *runs* and not to specific snapshots of training.

Don't you agree that it makes much more sense to ask the question: "How well does this model do on my validation set?" and "What is the best model I trained?" rather than to ask "How does the learning curve of the training run this model was trained with look like?" and "What was the best run I had?". By just associating a model with a run we enable people to answer the second type of question. To answer the first I need to associate a model with a specific metric and using the step when the metric and model were logged seem a natural way to do this.

I am sorry if all this sounds a lot like complaining to you, but it's really just meant as a suggestion for improvement. I know MLflow would be a much, much more valuable tool to our team if it allowed us to programmatically query and compare different models and programmatically deploy the best models from our training runs. 

On the MLflow website it already says that you can

"Query and compare runs using any data analysis tool of your choice, for example, pandas."

But at the end of the day no-one cares about what their best *run* was. You want to know what your best *model* was.

In any case, thanks for open-sourcing a tool like this and patiently listening to feedback from random strangers on the internet! 🚀you guys rock!

**Comment by timonbimon** [comment]
@dbczumar would you mind re-opening the issue? I feel like it is not resolved. :) 
would also be happy to open a new one if you prefer that, since I feel this is less about logging artifacts with steps, but more about how to connect metrics with models and how to get the best model from a run.

**Comment by dbczumar** [comment]
@timonbimon Absolutely! Reopening it now. If possible, can you provide a step-by-step walkthrough of your current model training and model selection workflow, as well as a walkthrough of your preferred workflow? It would be useful to see how associating a `step` with an artifact would help to simplify model selection for your use case.

**Comment by apurva-koti** [comment]
@timonbimon - any update on your use case?

**Comment by to266** [comment]
Stepping in for @timonbimon here:

The current workflow is more or less as follows:
 - Have a number of runs in parallel, each:
   - Logs all the usual training and validation metrics.
   - Stores a checkpoint after each epoch (we're running PBT with `ray`, so it's intrinsically necessary).
   - Checks if this particular epoch was the best seen so far, and if so stores the epoch number in a tag `best_step`. Also stores a tag `best_step_value` with the corresponding best metric value.
   - The above found best checkpoint is logged as an `artifact` with the `artifact_path` being the epoch number / training step.
 - A central engine then looks at all runs, for each gets the `best_step_value` tag, and compares. 
 - It then picks the best one, and gets the corresponding `best_step` tag from the run.
 - Constructs the corresponding artifact path from the `best_step`.
 - Finally the logged artifact for the globally best checkpoint is retrieved.

If artifacts were intrinsically connected to step, the step casting into `artifact_path` could be completely avoided.

**Comment by cereballo** [comment]
This is a feature I've been looking for as well! I hope the MLflow team considers it some more as it would be a great addition in my opinion :).

**Comment by Queuecumber** [comment]
I train image enhancement/restoration models and saving images per epoch is an extremely important feature for visualizing the training process. Also, while I appreciate mlflow's simplicity, all the major competitors (comet, wandb, neptune) have UI support for visualizing images (and videos, audios, texts, etc) per training step.
